### PR TITLE
Refactor ECS to use the Sparse Set data structure  

### LINKED ISSUE #16: Research Other ECS, make sure using right datatypes/implimentation
- best datatype for actual ECS https://www.geeksforgeeks.org/sparse-set/
- https://austinmorlan.com/posts/entity_component_system/
- https://github.com/skypjack/entt

### PR DESCRIPTION
**Pull request by AlecTroemel**
Closes #16 

the only incompatibility here is that I **removed** the ability for components to have single fields. So `(def-componet pizza :string)` is no more. The reason is that its impossible to update those fields, since they're often simple data types which are passed by value! This is actually broken in the `01-ecs.janet` example with the lives component.  

#### TODO:
- [X] get sparse set working on `examples/01-ecs.janet`  
- [x] cache view results. need to invalidate relevant cache whenever a component or entity is added/removed
- [x] either remove component limit, or find clean api to configure it. Should we try to reuse stale entity id numbers??
- [x] verify all other examples and code works 
- [x] create a minimal performance comparison

### CODE REVIEW

**Comment by pepe** [comment]
I am on the phone ATM and it looks good, but I can do a proper review tomorrow first thing in the office if you want to. Let me know, please

**Comment by AlecTroemel** [comment]
@pepe you're input is always appreciated! However I will say this MR is not quiet ready yet. Specifically, adding caching I suspect will change the code here up pretty significantly. I also am probably going to bite the bullet and write some tests, which could make reviewing easier?

All that to say, maybe hold off until I work through more of the checklist?

**Comment by pepe** [comment]
Oh. Thanks for the info! Maybe consider changing PR status to draft? May I ask, is the checklist visible somewhere?

**Comment by AlecTroemel** [comment]
hmm, the todo should be right in the description of this PR. I did edit the desc after creating the PR. I also just updated the title to include "DRAFT"

**Comment by AlecTroemel** [comment]
added a super simple caching of views results. Some super simple timing looks _very promising_ :smile: 

```clojure
(use /junk-drawer)

(def-component position :x :number :y :number)
(def-component velocity :x :number :y :number)

(def world (create-world))

(for i 0 10000
  (add-entity world
              (position :x i :y i)
              (velocity :x i :y i)))

(def-system a {es [:position]} nil)
(def-system b {es [:velocity]} nil)
(def-system c {es [:position :velocity]} nil)

(register-system world a)
(register-system world b)
(register-system world c)

(for i 0 1000
  (:update world 1))
```

timings 
```bash
# main branch
> time janet perf.janet
janet perf.janet  47.26s user 0.01s system 99% cpu 47.276 total

# Sparse set
> time janet perf.janet
janet perf.janet  0.06s user 0.00s system 99% cpu 0.063 total
```

However I still need to invalidate the cache on entity/component adding/removing. I doubt that will drastically change timingresults though!

**Comment by pepe** [comment]
It's a pity I blame the phone GH app 😂. Great work anyway, more from the table with display, a real keyboard, coffee, and glasses. So not to embarrass me even more 😉

File: junk-drawer/cache.janet [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+
+(defn cache-partial-clear [self component]
+  (loop [query :keys self
+         :when (index-of component query false)]
+    (put self query nil)))
+
+(defn init []
+  (table/setproto
+   @{}
+   @{:insert cache-insert
```
Comment (pepe) [inline junk-drawer/cache.janet]: I would remove the `cache-` prefix from the function names.

File: junk-drawer/ecs.janet [hunk 2]
```diff
@@ -1,92 +1,135 @@
+(import /junk-drawer/sparse-set :as sparse-set)
```
Comment (pepe) [inline junk-drawer/ecs.janet]: The `:as` is redundant here.

File: junk-drawer/ecs.janet [hunk 3]
```diff
@@ -1,92 +1,135 @@
...
-       ,;(map
-           |(quasiquote (put-in ,$db [,(keyword (first $)) ,$id] ,$))
-           components)
-       (put ,$wld :id-counter (inc ,$id))
-       ,$id)))
+           ,$cmp-name ,(keyword (first component))]
+       (when (nil? (get-in ,$wld [:database ,$cmp-name]))
+         (put-in ,$wld
+                  [:database ,$cmp-name]
+                  (sparse-set/init MAX_ENTITY_COUNT MAX_ENTITY_COUNT)))
```
Comment (pepe) [inline junk-drawer/ecs.janet]: I would unquote the `sparse-set/init` function here, which could solve some of the code shuffling in this PR. It also applies to some more functions in the quasi quotes down the road.

**Review by pepe — COMMENTED** [review]
Please take this as my point of view. The thing is, most of the recommendations are more of my way of doing things in Janet. Also, I do not have any Janet projects to review code in ATM :-D. I hope you will take this friendly with an open mind.

There are two more,  which are repeated in your code:

- I always try to format docstring as prose, so sentences start with capital letters and end with the period.
- I am always using `use` where I do not care about the prefix of the import. If for no other reason but the variable arity of the macro https://github.com/janet-lang/janet/blob/8f0a1ff/src/boot/boot.janet#L2948.

File: junk-drawer/cache.janet (hunk 1 above)
Comment (AlecTroemel) [reply junk-drawer/cache.janet]: Yaaa, I coded myself into a corner by naming one of the functions `get`, and I didn't want to override the standard API. The functions are small enough that I could probably get away with in-lining them right into the prototype object, but that kind of feels wrong also.

File: junk-drawer/ecs.janet (hunk 3 above)
Comment (AlecTroemel) [reply junk-drawer/ecs.janet]: I had a lot of problems with this macro, so I'll definitely try that out!

File: junk-drawer/cache.janet (hunk 1 above)
Comment (pepe) [reply junk-drawer/cache.janet]: That makes sense. If functions will not be part of the public API I am using good old _ at the start of the name in these cases, together with `defn-`. Which indeed does not feel great too

File: junk-drawer/ecs.janet (hunk 3 above)
Comment (pepe) [reply junk-drawer/ecs.janet]: It is one of the very good Janet  macro tricks

File: junk-drawer/cache.janet (hunk 1 above)
Comment (AlecTroemel) [reply junk-drawer/cache.janet]: ah ya totally forgot to use `defn-`, `_get` is good to me!

File: junk-drawer/ecs.janet [hunk 4]
```diff
@@ -1,92 +1,135 @@
...
-       ,;(map
-           |(quasiquote (put-in ,$db [,(keyword (first $)) ,$id] ,$))
-           components)
-       (put ,$wld :id-counter (inc ,$id))
-       ,$id)))
+           ,$cmp-name ,(keyword (first component))]
+       (when (nil? (get-in ,$wld [:database ,$cmp-name]))
+         (put-in ,$wld
+                  [:database ,$cmp-name]
+                  ,(sparse-set/init MAX_ENTITY_COUNT MAX_ENTITY_COUNT)))
```
Comment (pepe) [inline junk-drawer/ecs.janet]: I am sorry, but I meant to unquote the function, not the result of the call

Comment (AlecTroemel) [reply junk-drawer/ecs.janet]: you mean this? :exploding_head:

```clojure
(,sparse-set/init MAX_ENTITY_COUNT MAX_ENTITY_COUNT)
```

**Comment by AlecTroemel** [comment]
I've made a couple of updates here

- made sure all the docstrings start with capital letters and end with the period. I also reworded some things to hopefully be more clear.
- `use` is used where it can be, thanks to the `quote the function` deep magic
- added a `test/performance/stress-test.janet` script that has horrible macros but tests more parts of the ECS. Seems like there's a bottleneck creating/deleting entities from the world. However the difference in perf between this and main is still about the same

```bash
# main
janet test/performance/stress-test.janet  171.47s user 0.02s system 99% cpu 2:51.52 total

# sparse set
janet test/performance/stress-test.janet  2.90s user 0.02s system 99% cpu 2.923 total
```

File: test/performance/stress-test.janet [hunk 5]
```diff
@@ -0,0 +1,76 @@
+# run with 'time janet stress-test.janet'
+# NOTE: the macros here and incredibly unhygenic with probably a lot of accidental binding bapture!
+
+(use /junk-drawer)
+
+(var ALPHABET ["a" "b" "c" "d" "e" "f" "g" "h" "i"
```
Comment (pepe) [inline test/performance/stress-test.janet]: I guess this could be `def`ed as a constant?

File: test/performance/stress-test.janet [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+
+(defmacro def-component-alphabet []
+  "Define a component for every letter of the alphabet."
+  ~(upscope
+    ,;(map |(tuple ~def-component (symbol $) :val :number)
+           ALPHABET)))
+
+(defmacro def-systems-alphabet []
+  "Define and register a system with 2 queries, each with 2 alphabet components."
+  ~(upscope
```
Comment (pepe) [inline test/performance/stress-test.janet]: I have not tried it, I hope I will have time to try it later today, but I think `upscope`s are not needed here.

File: test/performance/stress-test.janet [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+    # define system
+    ,;(map (fn [i]
+             (tuple ~def-system (symbol "sys-" i)
```
Comment (pepe) [inline test/performance/stress-test.janet]: Single symbols are better `quote`d with `'`. There is no need to `quasiquote` them with `~`.

File: junk-drawer/timers.janet [hunk 8]
```diff
@@ -1,4 +1,4 @@
-(use ./ecs)
```
Comment (pepe) [inline junk-drawer/timers.janet:1]: This change goes against the stated move to `use`. May I ask what the reasons are? I am asking out of curiosity.

**Review by pepe — COMMENTED** [review]
I left some minor comments, yet as a whole, I am delighted with these many improvements to the core functionality. I will try to check the code running on my machine today. Thank you!

File: test/performance/stress-test.janet (hunk 5 above)
Comment (AlecTroemel) [reply test/performance/stress-test.janet]: resolving these comments, since you fix 'em https://github.com/AlecTroemel/junk-drawer/pull/22

File: junk-drawer/timers.janet (hunk 8 above)
Comment (AlecTroemel) [reply junk-drawer/timers.janet:1]: huh, I  thought I changed these back to `use` :facepalm: , must have forgot to commit it. I'll blame git

Comment (AlecTroemel) [reply junk-drawer/timers.janet:1]: imports back to `use`

**Comment by pepe** [comment]
Thank you very much for considering my codes!

The one thing I have been wondering about from the beginning of this adventure: What if we make sparse-set Janet abstract type in C. After this PR gets in and settled, would you consider checking that route?

**Comment by AlecTroemel** [comment]
yes, I've actually been thinking about that a lot also! 

One concern would be loosing any sort of portability. Consider if someone eventually added Janet to [TIC-80](https://github.com/nesbox/TIC-80/). With a pure Janet implementation you could copy/paste the code here and be on your way. But with C bindings I don't know if/how that would work. However that's a very specific concern for something that doesn't exist yet, and since right now I'm more interested in doing stuff with Jaylib/Raylib its definitely not a deal breaker.

Another thought would be if were writing C code, would it be a _ton more work_ to go the extra mile and bind to an existing ECS like FLECS?.. I'm not sure that even fits the philosophy of this code base, plus my gut says I'm underestimating the work there haha. 

TL;DR: Yes, I'm open to moving the sparse set, and maybe even the view/query logic, to C!

**Comment by pepe** [comment]
That is heavenly chords to my ears. I would even go one step further and write it in `jpm/cgen`.

As for wrapping some other code, I am not interested in this approach as I would like to stay in Janet's realm mentally

**Comment by AlecTroemel** [comment]
I plan on merging this today or tomorrow!
